### PR TITLE
Production ready client generation

### PR DESCRIPTION
**Pull request by Eunovo**
Generate only one file with all client documents as discussed [#1085 ](https://github.com/aerogear/graphback/issues/1085)

### CODE REVIEW

File: packages/graphback-cli/tests/workflow-test.ts [hunk 1]
```diff
@@ -30,36 +30,7 @@ test('Test cli workflow', async () => {
...
-  expect(
-    existsSync(join(basePath, "client/src/graphql/mutations/createComment.graphql"))
-  ).toBe(true);
-  expect(
-    existsSync(join(basePath, "client/src/graphql/mutations/updateComment.graphql"))
-  ).toBe(true);
-  expect(
-    existsSync(join(basePath, "client/src/graphql/mutations/updateNote.graphql"))
-  ).toBe(true);
+  expect(existsSync(join(basePath, "client/src/graphql/generated.graphql"))).toBe(true);
```
Comment (wtrocki) [inline packages/graphback-cli/tests/workflow-test.ts:33]: default name could be `graphback`

File: packages/graphback-codegen-client/src/helpers/writeDocuments.ts [hunk 2]
```diff
@@ -1,17 +1,31 @@
...
 
-export const writeDocumentsToFilesystem = (clientPath: string, generatedClient: ClientTemplates, extension: 'ts' | 'graphql') => {
-    if (!existsSync(clientPath)) {
-        mkdirSync(clientPath, { recursive: true })
-    }
-    Object.keys(generatedClient).forEach((folder: string) => {
-        const currentFolder = `${clientPath}/${folder}`
-        if (!existsSync(currentFolder)) {
-            mkdirSync(currentFolder)
+interface IWriter {
```
Comment (wtrocki) [inline packages/graphback-codegen-client/src/helpers/writeDocuments.ts:4]: Why we need this?

Comment (wtrocki) [reply packages/graphback-codegen-client/src/helpers/writeDocuments.ts:4]: This is always going to run on Node.js API so we can hardcode those

Comment (Eunovo) [reply packages/graphback-codegen-client/src/helpers/writeDocuments.ts:4]: So I can inject my own functions for testing

Comment (wtrocki) [reply packages/graphback-codegen-client/src/helpers/writeDocuments.ts:4]: Not needed. You can dissengage code that creates string from creation of the file. Creation of the file code do not require testing etc. For the rest we use snapshots effectively

**Comment by wtrocki** [comment]
Sample client app needs to be compiled with this changes. See `./templates` folder

**Comment by Eunovo** [comment]
> Sample client app needs to be compiled with this changes. See `./templates` folder

Templates use `./src/graphql/**/**.graphql` as document inputs. The input still match `./src/graphql/graphback.graphql`

**Comment by wtrocki** [comment]
Yep.. `yarn codegen needs to be run on the client to see if nothing changed`.
This will give us feedback if that approach works.

**Comment by wtrocki** [comment]
CC @machi1990 - This can also affect (hopefully positively OVP)

**Comment by machi1990** [comment]
> CC @machi1990 - This can also affect (hopefully positively OVP)

Will be part of 0.14.0?

**Comment by wtrocki** [comment]
Yes. Once this lands I will release new dev version. 
This will have huge impact on the changes. We no longer will need schema generation as well.

**Comment by Eunovo** [comment]
I need help with the `runtime-workflow` test in `integration`. Something might be wrong with my setup. My tests hang on this [line](https://github.com/aerogear/graphback/blob/master/integration/tests/runtime-workflow.ts#L43)

File: templates/ts-react-apollo-client/src/generated-types.tsx [hunk 3]
```diff
@@ -15,7 +15,7 @@ export type Scalars = {
 
 /**  @model  */
```
Comment (wtrocki) [inline templates/ts-react-apollo-client/src/generated-types.tsx:16]: I think our config should have enabled subscriptions for crud. Generated file has that disabled and that is why we see changes. Let's change all crud options in graphqlrc file and regenerate it again

**Comment by wtrocki** [comment]
Integration tests require running database. CI runs it. check docker compose for mongo to run it locally

File: templates/ts-react-apollo-client/src/generated-types.tsx (hunk 3 above)
Comment (Eunovo) [reply templates/ts-react-apollo-client/src/generated-types.tsx:16]: So `ts-react-apollo-client` does not have a `.graphqlrc.yml`. I used the config in `ts-apollo-postgres-backend`; this one doesn't enable subscriptions. The config in `ts-apollo-mongodb-backend` does enable subscriptions; Is this the config used to generate `ts-react-apollo-client` grapql documents?

**Comment by Eunovo** [comment]
> Integration tests require running database. CI runs it. check docker compose for mongo to run it locally. You might need to change ports in test

The integration tests uses Postgres with knex. I have a local Postgres server running and I still can't run the test. Knex only supports SQL databases so I can't use it with Mongo. I would have to change the code to use my mongodb server

**Comment by wtrocki** [comment]
My fault. you are right https://github.com/aerogear/graphback/blob/master/integration/tests/runtime-workflow.ts#L32

Now why tests are failing. Any error?

**Comment by Eunovo** [comment]
> Now why tests are failing. Any error?

I get three

`TypeError:  Cannot read property 'destroy' of undefined` [line 73](https://github.com/aerogear/graphback/blob/master/integration/tests/runtime-workflow.ts#L73)

and 

```
  console.log
    TypeError: pointerOptions.unixify is not a function
        at Array.collectGlob (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs\collect-sources.js:221:31)
        at next (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\utils\helpers.js:28:23)
        at Array.collectDocumentString (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs\collect-sources.js:218:5)
        at C:\Projects\graphback\node_modules\graphql-config\dist\core\src\utils\helpers.js:31:15
        at collectSources (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs\collect-sources.js:28:9)
        at processTicksAndRejections (internal/process/task_queues.js:97:5)
        at loadTypedefs (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs.js:12:21)
        at Object.loadSchema (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\schema.js:6:21)

      at tests/runtime-workflow.ts:65:13
          at Generator.throw (<anonymous>)


 console.error
    Unhandled error

      at process.uncaught (../node_modules/jest-jasmine2/build/jasmine/Env.js:248:21)

  console.error
    TypeError: pointerOptions.unixify is not a function
        at Array.collectGlob (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs\collect-sources.js:221:31)
        at next (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\utils\helpers.js:28:23)
        at Array.collectDocumentString (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs\collect-sources.js:218:5)
        at C:\Projects\graphback\node_modules\graphql-config\dist\core\src\utils\helpers.js:31:15
        at collectSources (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs\collect-sources.js:28:9)
        at processTicksAndRejections (internal/process/task_queues.js:97:5)
        at loadTypedefs (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\load-typedefs.js:12:21)
        at Object.loadSchema (C:\Projects\graphback\node_modules\graphql-config\dist\core\src\schema.js:6:21)

      at process.uncaught (../node_modules/jest-jasmine2/build/jasmine/Env.js:249:21)

Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```

**Comment by wtrocki** [comment]
Ok. This is kinda well known issue of the integration tests. When something is failing in before it will not print error at all. It is probably failing to connect to the database - usually port or username issue. If you run one from docker-compose in our template it has different port

**Comment by Eunovo** [comment]
@wtrocki  I was able to get the integration tests running.

There is a new problem. What do we do about `gqlwithfragment` format? It's no longer needed as all documents are now in one file. Currently, generating all documents in one file using this format causes duplicate fragments to be generated

**Comment by wtrocki** [comment]
I would not call it a problem. Do we need any other formats? What js the value of ts format?

**Comment by Eunovo** [comment]
> Do we need any other formats?

Apart from `ts`? I don't think so

> What js the value of ts format?

I don't understand this

Can I remove the `gqlwithfragment` format? It will require a change to docs.

**Comment by wtrocki** [comment]
But what is the value of the ts format when we have GraphQL one?

**Comment by Eunovo** [comment]
We don't need any format apart from graphql; since we will use graphql-codegen to generate the ts types.

**Comment by Eunovo** [comment]
I tried to rebase and ended up messing the branch.
@wtrocki  can you merge it like this or should I create a new branch.

**Comment by Eunovo** [comment]
I see a comment about generating `ts` for backwards compatibility [here](https://github.com/aerogear/graphback/blob/master/packages/graphback-codegen-client/src/ClientCRUDPlugin.ts#L14).

Should I leave the `gqlwithfragment` format for backwards compatibility too.

**Comment by wtrocki** [comment]
I think the simplest then better.

We can have `ts` and `graphql` format with the single file save. 
GraphQLWithFragment makes no sense.

**Comment by wtrocki** [comment]
Requires rebase

File: templates/ts-react-apollo-client/src/generated-types.tsx (hunk 3 above)
Comment (wtrocki) [reply templates/ts-react-apollo-client/src/generated-types.tsx:16]: Yes.. but we should enable all actually. I think they were reverted or something. We should enable all CRUD flags in this PR. Without that we can't really test it properly. Also types here were generated with all crud enabled.

**Comment by wtrocki** [comment]
Let's also some single page sentence to release-notes.md

**Comment by wtrocki** [comment]
We should enable all CRUD flags in this PR. Without that we can't really test it properly. Also types here were generated with all crud enabled.

**Comment by wtrocki** [comment]
Seee https://github.com/aerogear/graphback/blob/master/templates/ts-apollo-postgres-backend/.graphqlrc.yml#L13-L16

**Comment by Eunovo** [comment]
@wtrocki  review before I rebase

**Comment by wtrocki** [comment]
Not sure what you mean. Do you mean me chekcing not rebased PR.I normally do review on rebased code that has all issues resolved

**Comment by wtrocki** [comment]
I see that we added ts format. We should keep grahql format and the same model.

**Comment by Eunovo** [comment]
> Not sure what you mean. Do you mean me chekcing not rebased PR.I normally do review on rebased code that has all issues resolved

ok

**Comment by Eunovo** [comment]
> I see that we added ts format. We should keep grahql format and the same model.

Not sure what you mean. Should I remove the ts format? I thought you need to verify that it works

**Comment by wtrocki** [comment]
Yes.. But we do not want this on master as we use codegen

**Comment by Eunovo** [comment]
> Were you able to run client app?

Yes.

**Comment by Eunovo** [comment]
Ready for review

**Comment by wtrocki** [comment]
Looks good. Have you tried to run sample app with backend?

**Comment by Eunovo** [comment]
Yes. I ran the sample app. It worked perfectly.

**Comment by machi1990** [comment]
@Eunovo @wtrocki I gave this a shot here https://github.com/aerogear/OpenVolunteerPlatform/pull/93 and it work nicely. This is really good work. 

I have a couple of comments:

- While the upgrade will work out of the box, the old generated gql query / fragment files will be there, dangling in user source codes (see the amount of deleted files in the PR I linked).  I'd have loved to see a good message in the console indicating which files can be deleted but a mention in the release note / docs that crud files can be deleted.

- It seems to me that `gqlwithfragment` is no longer supported https://github.com/aerogear/graphback/pull/1297/files#diff-68c63497d04c1e9533c6b395ed2a788fR31 but docs on the master do not reflect that https://github.com/aerogear/graphback/blob/06ea54ebd4650586a6d20830b467fdf687ddd676/docs/plugins/graphback-client.md#plugin-config

**Comment by wtrocki** [comment]
Actually we have zero info in release notes and docs
@Eunovo Can you help with this?

> graphback.(graphql | ts)

is name hardcoded? I got that we will get option to control what name will be used and default will be graphback or generated.graphql/ts

**Comment by machi1990** [comment]
> Actually we have zero info in release notes and docs
> @Eunovo Can you help with this?

This will be awesome.
 
> > graphback.(graphql | ts)
> 
> is name hardcoded? I got that we will get option to control what name will be used and default will be graphback or generated.graphql/ts

Yes, it is hardcoded in the meantime https://github.com/aerogear/graphback/pull/1297/files#diff-a4d6c30af7374831c26dbbed7db947c6R15, having an option to change it will be welcomed.

**Comment by Eunovo** [comment]
> Actually we have zero info in release notes and docs
> @Eunovo Can you help with this?
> 
> > graphback.(graphql | ts)
> 
> is name hardcoded? I got that we will get option to control what name will be used and default will be graphback or generated.graphql/ts

@wtrocki I can change the code to use the outputPath so that generated code if
`outputPath = "src/graphql(.ts | .graphql | no extension)"` generated code will be in `src/graphql(ts | graphql)`

**Comment by wtrocki** [comment]
Yep. And documentation should be updated as well

**Pull request merged**